// mongo-gridfs/dist/mongo-gridfs.class.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var bson_1 = require("bson");
var fs = require("fs");
var mongodb_1 = require("mongodb");
var osTmpdir = require("os-tmpdir");
var uniqueFilename = require("unique-filename");
var MongoGridFS = /** @class */ (function () {
    /**
     * Constructor
     * @param {connection} connection
     * @param {string} bucketName
     */
    function MongoGridFS(connection, bucketName) {
        if (bucketName === void 0) { bucketName = 'fs'; }
        this.connection = connection;
        this.bucketName = bucketName;
    }
    Object.defineProperty(MongoGridFS.prototype, "bucket", {
        get: function () {
            return new mongodb_1.GridFSBucket(this.connection, { bucketName: this.bucketName });
        },
        enumerable: true,
        configurable: true
    });
    MongoGridFS.getDownloadPath = function (object, options) {
        if (options === void 0) { options = {
            filename: false,
        }; }
        var finalPath = '';
        if (!options.targetDir) {
            if (typeof options.filename === 'string') {
                finalPath = osTmpdir() + "/" + options.filename;
            }
            else {
                if (options.filename === true) {
                    finalPath = osTmpdir() + "/" + object._id;
                }
                else {
                    finalPath = uniqueFilename(osTmpdir());
                }
            }
        }
        else {
            if (typeof options.filename === 'string') {
                finalPath = options.targetDir + "/" + options.filename;
            }
            else {
                if (options.filename === true) {
                    finalPath = object.filename;
                }
                else {
                    finalPath = uniqueFilename(options.targetDir);
                }
            }
        }
        return finalPath;
    };
    /**
     * Returns a stream of a file from the GridFS.
     * @param {string} id
     * @return {Promise<GridFSBucketReadStream>}
     */
    MongoGridFS.prototype.readFileStream = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            var object;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.findById(id)];
                    case 1:
                        object = _a.sent();
                        return [2 /*return*/, this.bucket.openDownloadStream(object._id)];
                }
            });
        });
    };
    /**
     * Save the File from the GridFs to the filesystem and get the Path back
     * @param {string} id
     * @param {IDownloadOptions} options
     * @return {Promise<string>}
     */
    MongoGridFS.prototype.downloadFile = function (id, options) {
        return __awaiter(this, void 0, void 0, function () {
            var object, downloadPath;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.findById(id)];
                    case 1:
                        object = _a.sent();
                        downloadPath = MongoGridFS.getDownloadPath(object, options);
                        return [2 /*return*/, new Promise(function (resolve, reject) { return __awaiter(_this, void 0, void 0, function () {
                                var _this = this;
                                return __generator(this, function (_a) {
                                    this.bucket.openDownloadStream(object._id)
                                        .once('error', function (error) { return __awaiter(_this, void 0, void 0, function () {
                                        return __generator(this, function (_a) {
                                            reject(error);
                                            return [2 /*return*/];
                                        });
                                    }); })
                                        .once('end', function () { return __awaiter(_this, void 0, void 0, function () {
                                        return __generator(this, function (_a) {
                                            resolve(downloadPath);
                                            return [2 /*return*/];
                                        });
                                    }); })
                                        .pipe(fs.createWriteStream(downloadPath, {}));
                                    return [2 /*return*/];
                                });
                            }); })];
                }
            });
        });
    };
    /**
     * Find a single object by id
     * @param {string} id
     * @return {Promise<IGridFSObject>}
     */
    MongoGridFS.prototype.findById = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.findOne({ _id: new bson_1.ObjectID(id) })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Find a single object by condition
     * @param filter
     * @return {Promise<IGridFSObject>}
     */
    MongoGridFS.prototype.findOne = function (filter) {
        return __awaiter(this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.find(filter)];
                    case 1:
                        result = _a.sent();
                        if (result.length === 0) {
                            throw new Error('No Object found');
                        }
                        return [2 /*return*/, result[0]];
                }
            });
        });
    };
    /**
     * Find a list of object by condition
     * @param filter
     * @return {Promise<IGridFSObject[]>}
     */
    MongoGridFS.prototype.find = function (filter) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.bucket.find(filter).toArray()];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Find objects by condition
     * @param stream
     * @param options
     */
    MongoGridFS.prototype.writeFileStream = function (stream, options) {
        var _this = this;
        return new Promise(function (resolve, reject) { return stream
            .pipe(_this.bucket.openUploadStream(options.filename, {
            aliases: options.aliases,
            chunkSizeBytes: options.chunkSizeBytes,
            contentType: options.contentType,
            metadata: options.metadata,
        }))
            .on('error', function (err) { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                reject(err);
                return [2 /*return*/];
            });
        }); })
            .on('finish', function (item) { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                resolve(item);
                return [2 /*return*/];
            });
        }); }); });
    };
    /**
     * Upload a file directly from a fs Path
     * @param {string} uploadFilePath
     * @param {IGridFSWriteOption} options
     * @param {boolean} deleteFile
     * @return {Promise<IGridFSObject>}
     */
    MongoGridFS.prototype.uploadFile = function (uploadFilePath, options, deleteFile) {
        if (deleteFile === void 0) { deleteFile = true; }
        return __awaiter(this, void 0, void 0, function () {
            var tryDeleteFile;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!fs.existsSync(uploadFilePath)) {
                            throw new Error('File not found');
                        }
                        tryDeleteFile = function (obj) {
                            if (fs.existsSync(uploadFilePath) && deleteFile === true) {
                                fs.unlinkSync(uploadFilePath);
                            }
                            return obj;
                        };
                        return [4 /*yield*/, this.writeFileStream(fs.createReadStream(uploadFilePath), options)
                                .then(tryDeleteFile)
                                .catch(function (err) {
                                tryDeleteFile();
                                throw err;
                            })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Delete an File from the GridFS
     * @param {string} id
     * @return {Promise<boolean>}
     */
    MongoGridFS.prototype.delete = function (id) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.bucket.delete(new bson_1.ObjectID(id), (function (err) { return __awaiter(_this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    if (err) {
                        reject(err);
                    }
                    resolve(true);
                    return [2 /*return*/];
                });
            }); }));
        });
    };
    return MongoGridFS;
}());
exports.MongoGridFS = MongoGridFS;
